Lock names include attributes values

LucasG0 opened the pull request:
IFC-1874

This PR makes the current locks more granular, ie instead of locking only on object kind, we also lock on values of the attributes contained within uniqueness constraint. In this way, we allow concurrent creation/updates of objects having different values on these key attributes. We cannot do it for relationships at it would require fetching relationships peers before locking.

EDIT 2025-09-30: this PR also contains fixes for the resource managers to prevent duplicate allocations. Locking is now done during the mutation but also during the allocation of the resource, thus making the locks "re-entrant" was required to avoid deadlocks within a same request.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Option to skip resource-pool allocation during object create/update.
  * Unified constraint-and-save operation that runs under managed multi-locks.

* **Refactor**
  * Preview-before-persist flow to derive validation/save fields and locks before saving.
  * Centralized lock-name computation and multi-lock orchestration across mutations.
  * Exposed process_pools flag to control pool handling.

* **Bug Fixes**
  * Branch-aware locking and improved reentrant lock behavior.

* **Tests**
  * Added tests for lock-name generation and reentrant lock behavior.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

LucasG0 committed:
Lock names include attributes values

ajtmccarty commented on backend/infrahub/lock_utils.py:
usually `utils` means that a file contains a bunch of things that don't really belong together and I don't think that is the case here. I think this file can be `locks.py` or even `lock_getter.py`

ajtmccarty commented on backend/infrahub/lock_utils.py:
```diff
@@ -0,0 +1,107 @@
+    return False
+
+
+def _get_lock_names_on_object_mutation(node: Node, branch: Branch, schema_branch: SchemaBranch) -> list[str]:
```
at least one of these should be a public method and I think it's this one

LucasG0 replied:
Right

ajtmccarty reviewed (commented):
could you please add a PR description that explains why the changes in `_get_kind_lock_names_on_object_mutation` will help with the locking slowness?

fatih-acar commented:
Quick remark: we should be very careful with user inputs, in this case you are lucky that Redis keys are very permissive regarding character set. I'm not sure it would be the case with all implementations.

We can instead of using the attribute value directly, use some kind of hex hash of it.

Edit: not saying we should do it right now, I don't even know performance implications of such a change.

LucasG0 commented:
@fatih-acar good point, I'll update to use the hash, I don't expect having performances issues on that, hashing a 1 million length string takes less than 1ms.

LucasG0 committed:
minor changes + hash attribute values

LucasG0 committed:
add checks for empty attr

LucasG0 committed:
use hash 2562

fatih-acar commented:
I think there's an issue with objects depending on a resource manager such as NumberPool... I end up with duplicate numbers allocated.

I think we still require to lock exclusively on kind when the object uses resource manager...

Edit: this issue is NOT related to https://github.com/opsmill/infrahub/pull/6866

fatih-acar commented:
> I think there's an issue with objects depending on a resource manager such as NumberPool... I end up with duplicate numbers allocated.
> 
> I think we still require to lock exclusively on kind when the object uses resource manager...
> 
> Edit: this issue is NOT related to #6866

https://github.com/opsmill/infrahub/pull/7025 is not enough to fix the issue because the `get_resource()` call is done within a DB transaction when we're inside the Create/Upsert mutation path...
One solution to fix this would be to apply the same filter we do on uniqueness constraints, but also on attributes that depend on a resource manager.

fatih-acar committed:
Revert "Lock on names for GraphQLQueryGroup (#6899)"

This reverts commit fca14f24770dca3a8a759fcc13c551b295d925e5.

fatih-acar committed:
fix: make sure to lock on resource manager allocations

We should also lock regardless of the branch when allocating from a
resource manager.

Signed-off-by: Fatih Acar <fatih@opsmill.com>

fatih-acar committed:
Revert "fix: make sure to lock on resource manager allocations"

BROKEN: this does not handle double lock/deadlock issues when we acquire
a resource pool lock in run_constraints_and_save() and then in
get_resource(). We probably require a mechanism to check if the lock is
actually held by the current worker, if so it can proceed without
acquiring the lock.

This reverts commit 31d79407eb8e0111c024693d7c3a7cec66e8bde2.

fatih-acar committed:
Reapply "fix: make sure to lock on resource manager allocations"

Make locks re-entrant within the same asyncio context.

Signed-off-by: Fatih Acar <fatih@opsmill.com>

fatih-acar committed:
fix(backend): lock before beginning transactions

WIP: still need to fix the uniqueness constraint lock on update.

Signed-off-by: Fatih Acar <fatih@opsmill.com>

fatih-acar committed:
fix(backend): lock on resource manager attributes

Signed-off-by: Fatih Acar <fatih@opsmill.com>

fatih-acar committed:
fix(backend): fix resource pool allocation in update mutation

Signed-off-by: Fatih Acar <fatih@opsmill.com>

fatih-acar committed:
fix(backend): ensure ipam mutations lock before transaction

Fixes "Lock names include attributes values"

Signed-off-by: Fatih Acar <fatih@opsmill.com>

fatih-acar commented:
Fixes https://github.com/opsmill/infrahub/issues/7254

fatih-acar commented:
We can probably re-enable locking within branches now that they are more granular?

fatih-acar committed:
fix(backend): use contextvars for reentrant locking

Previous implementation was wrong due to asyncio context switching.

Signed-off-by: Fatih Acar <fatih@opsmill.com>

fatih-acar committed:
fix(backend): lock on mutations for branches

We can enable locking for consistency on branches now that they are more
granular and avoid performance penalty.

Signed-off-by: Fatih Acar <fatih@opsmill.com>

fatih-acar committed:
fix(backend): coderabbit fix

Signed-off-by: Fatih Acar <fatih@opsmill.com>

ajtmccarty commented on backend/infrahub/lock_getter.py:
```diff
@@ -0,0 +1,124 @@
+
+def get_lock_names_on_object_mutation(node: Node, schema_branch: SchemaBranch) -> list[str]:
+    """
+    Return lock names for object on which we want to avoid concurrent mutation (create/update). Except for some specific kinds,
+    concurrent mutations are only allowed on non-main branch as objects validations will be performed at least when merging in main branch.
+    Lock names include kind, some generic kinds, and values of attributes of corresponding uniqueness constraints.
+    """
```
pretty sure this is leftover from months ago, but I don't think this is accurate any longer
also, I don't think this docstring belongs here as it references logic way outside of its scope

fatih-acar replied:
Indeed, let me know what you think of the new version

ajtmccarty commented on backend/infrahub/lock_getter.py:
```diff
@@ -0,0 +1,124 @@
+            lock_names.append(lock_name)
+
+    # Check if node is using resource manager allocation via attributes
+    for attr_name in getattr(node, "_attributes", []):
```
we probably do this in many other places, but I think the more correct way to get these names is using the schema linked to the `node` object
```python
for attr_name in node.get_schema().attribute_names:
```

fatih-acar replied:
Thanks, that's the type of thing I'm not aware of :)

ajtmccarty commented on backend/infrahub/lock_getter.py:
```diff
@@ -0,0 +1,124 @@
+    def _add_pool_lock(pool_id: str) -> None:
+        lock_name = f"resource_pool.{pool_id}"
+        if lock_name not in lock_names:
+            lock_names.append(lock_name)
```
if `lock_names` was a `set` you could probably remove this `_add_pool_lock` function-in-a-function

fatih-acar replied:
Smarter than AI :p

ajtmccarty commented on backend/infrahub/lock_getter.py:
```diff
@@ -0,0 +1,124 @@
+            _add_pool_lock(attribute.from_pool["id"])
+
+    # Check if relationships allocate resources
+    for rel_name in node._relationships:
+        rel_manager: RelationshipManager = getattr(node, rel_name)
+        for rel in rel_manager._relationships:
```
maybe we need a `Node.get_relationship_managers` method and a `RelationshipManager.get_fetched_relationships` method to avoid accessing these private variables outside of the classes

fatih-acar replied:
Yes I believe we do this in many places to access relationships without fetching them. I agree on this point but we can probably do that in another PR?

ajtmccarty commented on backend/infrahub/lock_getter.py:
```diff
@@ -0,0 +1,124 @@
+
+    lock_kinds = _get_kinds_to_lock_on_object_mutation(node.get_kind(), schema_branch)
+    for kind in lock_kinds:
+        schema = schema_branch.get(name=kind)
```
`duplicate=False`

fatih-acar replied:
Good catch!

ajtmccarty commented on backend/infrahub/lock_getter.py:
```diff
@@ -0,0 +1,124 @@
+
+
+def _hash(value: str) -> str:
+    # Do not use builtin `hash` for lock names as due to randomization results would differ between
+    # different processes.
```
I learned something new today

fatih-acar replied:
Thanks coderabbit :)

ajtmccarty commented on backend/infrahub/lock.py:
is this all to support the "re-entrant" locks you mentioned in the description? meaning a lock that can be acquired multiple times?
how does this prevent different workers from acquiring the same lock if the lock can be acquired more than once?

fatih-acar replied:
Yes exactly
We track this behavior using a counter within a Contextvar: this is to have a asyncio context scoped counter (this is similar to TLS (thread local storage) but for asyncio coroutines, see https://peps.python.org/pep-0567/)

Different worker is necessarily within another coroutine, so in this case since the refcount is 0, it will try to acquire the global lock, which will be held by another worker.
A different coroutine within the same worker will also try to acquire the lock (in case of a concurrent graphql mutation running on the same worker for example).

fatih-acar replied:
The main goal is just to fix the following call stack:
Create mutation holds resource pool lock -> create new node internally using Node.new() -> triggers resource pool get_resource() -> get_resource locks again using the same resource pool id -> deadlock

We introduced `get_resource locks again using the same resource pool id` in https://github.com/opsmill/infrahub/pull/7025 to lock direct mutation calls to the resource pools...
A cleaner solution would be to move the lock from the `get_resource` call up to the different top-level mutation handlers.

ajtmccarty commented on backend/infrahub/graphql/mutations/ipam.py:
it would be nice to reduce some of the duplicated code in here, but that is seems out of scope for this one

fatih-acar replied:
Agree

ajtmccarty commented on backend/infrahub/graphql/mutations/main.py:
```diff
@@ -220,42 +210,52 @@ async def _call_mutate_update(
         Wrapper around mutate_update to potentially activate locking and call it within a database transaction.
         """
+        before_mutate_profile_ids = await get_profile_ids(db=db, obj=obj)
+        fields_to_validate = list(data)
+        fields = list(data.keys())
```
are these different?

fatih-acar replied:
I believe `fields_to_validate` is used for the constraints checker, this is what I understand from the previous state of the code. So yes they are different and required?

ajtmccarty commented on backend/infrahub/graphql/mutations/main.py:
```diff
@@ -220,42 +210,52 @@ async def _call_mutate_update(
+            id=obj.get_id(),
+            branch=branch,
+            include_owner=True,
+            include_source=True,
```
are owner and source necessary for this?

fatih-acar replied:
Indeed good catch! AI not smart enough

ajtmccarty reviewed (commented):
some questions and comments (mostly on things you did not write)
and a couple more questions
1. is it possible to add any sort of graphql-level tests that actually cover the locks as they will be used? seems like it would be hard, but maybe it's possible by creating/updating overlapping objects and checking timestamps of the changes or something
2. is this change big enough to be on develop instead of stable?

fatih-acar committed:
post review fixes

Signed-off-by: Fatih Acar <fatih@opsmill.com>

fatih-acar commented:
Thanks for reviewing @ajtmccarty 

> some questions and comments (mostly on things you did not write) and a couple more questions
> 
> 1. is it possible to add any sort of graphql-level tests that actually cover the locks as they will be used? seems like it would be hard, but maybe it's possible by creating/updating overlapping objects and checking timestamps of the changes or something
> 2. is this change big enough to be on develop instead of stable?

1. I think yes, but it would be a bit hard as you wonder. It's pretty easy to stress test this using our internal scale testing suite since it massively creates objects concurrently (this is how I caught most of the issues that are resolved in this PR).
2. I think this should go in develop for the 1.6 release, but we might need to cherry-pick some changes to fix https://github.com/opsmill/infrahub/issues/7254 in stable...

fatih-acar committed:
fix(backend): disable metrics for mutation locks

Granular locking produces more cardinality for lock names thus bloating
prometheus metrics leading to high memory usage.
Completely disable metrics for these locks as they would be useless for
the moment.

Signed-off-by: Fatih Acar <fatih@opsmill.com>

fatih-acar commented:
Replaced by https://github.com/opsmill/infrahub/pull/7508
